File: desktop/src/main/kotlin/com/videoroom/data/UpdateChecker.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 VideoRoom Contributors

package com.videoroom.data

import com.videoroom.AppVersion
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.URL

data class ReleaseInfo(
    val version: String,
    val releaseUrl: String,
    val tagName: String
)

object UpdateChecker {

    suspend fun checkLatestRelease(owner: String, repo: String): ReleaseInfo? =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("https://api.github.com/repos/$owner/$repo/releases/latest")
                val conn = url.openConnection() as HttpURLConnection
                conn.setRequestProperty("Accept", "application/vnd.github+json")
                conn.setRequestProperty("User-Agent", "VideoRoom/${AppVersion.CURRENT}")
                conn.connectTimeout = 8_000
                conn.readTimeout = 8_000
                if (conn.responseCode != 200) return@withContext null
                val body = conn.inputStream.bufferedReader().readText()
                val tagName = """"tag_name"\s*:\s*"([^"]+)"""".toRegex()
                    .find(body)?.groupValues?.getOrNull(1) ?: return@withContext null
                // html_url for the release page appears multiple times; the first occurrence is the release itself
                val htmlUrl = """"html_url"\s*:\s*"([^"]+)"""".toRegex()
                    .find(body)?.groupValues?.getOrNull(1) ?: return@withContext null
                val version = tagName.trimStart('v')
                ReleaseInfo(version = version, releaseUrl = htmlUrl, tagName = tagName)
            } catch (_: Exception) {
                null
            }
        }

    /** Returns true when [latestVersion] is strictly newer than [currentVersion] (semver comparison). */
    fun isNewer(latestVersion: String, currentVersion: String): Boolean {
        fun parts(v: String) = v.split('.').map { it.toIntOrNull() ?: 0 }
        val latest = parts(latestVersion)
        val current = parts(currentVersion)
        val len = maxOf(latest.size, current.size)
        for (i in 0 until len) {
            val l = latest.getOrElse(i) { 0 }
            val c = current.getOrElse(i) { 0 }
            if (l > c) return true
            if (l < c) return false
        }
        return false
    }
}

File: macos/ReelVault/Utilities/UpdateChecker.swift
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 ReelVault Contributors

import Foundation

struct ReleaseInfo {
    let version: String
    let releaseUrl: String
    let tagName: String
}

enum UpdateChecker {

    static func checkLatestRelease(owner: String, repo: String) async -> ReleaseInfo? {
        guard let url = URL(string: "https://api.github.com/repos/\(owner)/\(repo)/releases/latest")
        else { return nil }

        var request = URLRequest(url: url, timeoutInterval: 10)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("ReelVault/\(AppVersion.current)", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let tagName  = json["tag_name"]  as? String,
                  let htmlUrl  = json["html_url"]  as? String
            else { return nil }
            let version = tagName.hasPrefix("v") ? String(tagName.dropFirst()) : tagName
            return ReleaseInfo(version: version, releaseUrl: htmlUrl, tagName: tagName)
        } catch {
            return nil
        }
    }

    /// Returns true when `latest` is strictly newer than `current` (semver comparison).
    static func isNewer(_ latest: String, than current: String) -> Bool {
        func parts(_ v: String) -> [Int] { v.split(separator: ".").map { Int($0) ?? 0 } }
        let l = parts(latest), c = parts(current)
        let len = max(l.count, c.count)
        for i in 0..<len {
            let lv = i < l.count ? l[i] : 0
            let cv = i < c.count ? c[i] : 0
            if lv > cv { return true }
            if lv < cv { return false }
        }
        return false
    }
}
